### PR TITLE
Introduce parameter to restart l1->l2 messaging

### LINKED ISSUE #602: `flush` will load messages from previous tests
I've explained the situation in the [`README.md`](https://github.com/pscott/flush_example/blob/main/README.md) of this example repo: https://github.com/pscott/flush_example .

But basically, `flush` will load messages from L1 even if we have reset `L2` in between. This leads to harder testing, as one must either re-deploy contracts, or load L1 from a dump file (currently not possible using Anvil, as it's not a dump on request but a dump on exit).

### PR DESCRIPTION
**Pull request by FabijanC**
## Usage related changes

- Close #602
- Introduce optional parameter `restart_messaging_to_l2` to the `devnet_restart` RPC method. Behavior explained in the docs.

## Development related changes

- Extract `deploy_l2_msg_contract` from `setup_devnet` in messaging tests.

## Checklist:

- [x] Checked out the [contribution guidelines](CONTRIBUTING.md)
- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] No unused dependencies - `./scripts/check_unused_deps.sh`
- [x] No spelling errors - `./scripts/check_spelling.sh`
- [x] Performed code self-review
- [x] Rebased to the latest commit of the target branch (or merged it into my branch)
    -   Once you make the PR reviewable, please avoid force-pushing
- [x] Updated the docs if needed - `./website/README.md`
- [x] Linked the [issues](https://github.com/0xSpaceShard/starknet-devnet-rs/issues) resolvable by this PR - [linking info](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- [x] Updated the tests if needed; all passing - [execution info](https://github.com/0xSpaceShard/starknet-devnet-rs/blob/main/.github/CONTRIBUTING.md#test-execution)

### CODE REVIEW

File: crates/starknet-devnet/tests/test_messaging.rs [hunk 1]
```diff
@@ -660,4 +669,117 @@ mod test_messaging {
...
+        assert_eq!(user_balance_eth, 1.into()); // 2 - 1
+        assert_eq!(get_balance(&devnet, sn_l1l2_contract, user_sn).await, [Felt::ZERO]);
+
+        // Flush messages to have MessageToL2 executed.
+        devnet.send_custom_rpc("devnet_postmanFlush", json!({})).await.unwrap();
+
+        assert_eq!(user_balance_eth, deposit_amount.into());
+        assert_eq!(get_balance(&devnet, sn_l1l2_contract, user_sn).await, [Felt::ONE]);
+
+        devnet.send_custom_rpc("devnet_postmanFlush", json!({})).await.unwrap();
```
Comment (marioiordanov) [inline crates/starknet-devnet/tests/test_messaging.rs:755]: why this second flush is needed?

File: crates/starknet-devnet/tests/test_messaging.rs [hunk 2]
```diff
@@ -660,4 +669,117 @@ mod test_messaging {
...
+
+        let user_balance_eth = anvil.get_balance_l1l2(eth_l1l2_address, user_eth).await.unwrap();
+
+        assert_eq!(user_balance_eth, 1.into()); // 2 - 1
+        assert_eq!(get_balance(&devnet, sn_l1l2_contract, user_sn).await, [Felt::ZERO]);
+
+        // Flush messages to have MessageToL2 executed.
+        devnet.send_custom_rpc("devnet_postmanFlush", json!({})).await.unwrap();
+
+        assert_eq!(user_balance_eth, deposit_amount.into());
```
Comment (marioiordanov) [inline crates/starknet-devnet/tests/test_messaging.rs:752]: Is this check really needed? on this line https://github.com/0xSpaceShard/starknet-devnet-rs/blob/92dfd8b2ad46184e31cdfd2765a995d8b127aa1a/crates/starknet-devnet/tests/test_messaging.rs it is already checked

**Review by marioiordanov — CHANGES_REQUESTED** [review]

File: crates/starknet-devnet/tests/test_messaging.rs (hunk 2 above)
Comment (FabijanC) [reply crates/starknet-devnet/tests/test_messaging.rs:752]: Was missing a line which reads the balance: https://github.com/0xSpaceShard/starknet-devnet-rs/pull/609/commits/911e8b9bd056a109e0a189183743f74370b8ca3f

File: crates/starknet-devnet/tests/test_messaging.rs (hunk 1 above)
Comment (FabijanC) [reply crates/starknet-devnet/tests/test_messaging.rs]: I was testing redundant flushing while developing to see the behavior, but I guess it's not needed anymore, so I removed the line: https://github.com/0xSpaceShard/starknet-devnet-rs/pull/609/commits/911e8b9bd056a109e0a189183743f74370b8ca3f

**Review by marioiordanov — APPROVED** [review]

**Pull request merged**